src/rules/no_unused_vars.rs

```diff
@@ -0,0 +1,2052 @@
+    assert_lint_err::<NoUnusedVars>(
+      "export default (a, b) => { console.log(a); };",
+      19,
+    );
+    assert_lint_err::<NoUnusedVars>("try{}catch(err){};", 11);
+  }
+
+  #[test]
+  #[ignore = "control flow analysis is not implemented yet"]
+  fn no_unused_vars_err_assign_expr() {
+    assert_lint_err::<NoUnusedVars>("var a = 0; a = a + 1;", 0);
+    assert_lint_err::<NoUnusedVars>("var a = 0; a = a + a;", 0);
+    assert_lint_err::<NoUnusedVars>("var a = 0; a += a + 1", 0);
+    assert_lint_err::<NoUnusedVars>("var a = 0; a++;", 0);
+    assert_lint_err::<NoUnusedVars>("function foo(a) { a = a + 1 } foo();", 0);
+    assert_lint_err::<NoUnusedVars>("function foo(a) { a += a + 1 } foo();", 0);
+    assert_lint_err::<NoUnusedVars>("function foo(a) { a++ } foo();", 0);
+    assert_lint_err::<NoUnusedVars>("var a = 3; a = a * 5 + 6;", 0);
+    assert_lint_err::<NoUnusedVars>("var a = 2, b = 4; a = a * 2 + b;", 0);
+
+    assert_lint_err::<NoUnusedVars>("const a = 1; a += 1;", 6);
+  }
+
+  #[test]
+  #[ignore = "control flow analysis is not implemented yet"]
+  fn no_unused_vars_err_assign_to_self() {
+    assert_lint_err::<NoUnusedVars>("function foo(cb) { cb = function(a) { cb(1 + a); }; bar(not_cb); } foo();", 0);
+    assert_lint_err::<NoUnusedVars>(
+      "function foo(cb) { cb = function(a) { return cb(1 + a); }(); } foo();",
+      0,
+    );
+    assert_lint_err::<NoUnusedVars>(
+      "function foo(cb) { cb = (function(a) { cb(1 + a); }, cb); } foo();",
+      0,
+    );
+    assert_lint_err::<NoUnusedVars>(
+      "function foo(cb) { cb = (0, function(a) { cb(1 + a); }); } foo();",
+      0,
+    );
+  }
+
+  #[test]
+  fn no_unused_vars_err_10() {
+    assert_lint_err::<NoUnusedVars>(
+      "(function ({ a }, b ) { return b; })();",
+      13,
+    );
+    assert_lint_err_n::<NoUnusedVars>(
+      "(function ({ a }, { b, c } ) { return b; })();",
+      vec![13, 23],
+    );
+    assert_lint_err::<NoUnusedVars>(
+      "(function ([ a ], b ) { return b; })();",
+      13,
+    );
+    assert_lint_err_n::<NoUnusedVars>(
+      "(function ([ a ], [ b, c ] ) { return b; })();",
+      vec![13, 23],
+    );
+  }
+
+  #[test]
+  fn no_unused_vars_err_11() {
+    assert_lint_err::<NoUnusedVars>("var a = function() { a(); };", 4);
+    assert_lint_err::<NoUnusedVars>(
+      "var a = function(){ return function() { a(); } };",
+      4,
+    );
+    assert_lint_err::<NoUnusedVars>("const a = () => { a(); };", 6);
+    assert_lint_err::<NoUnusedVars>("const a = () => () => { a(); };", 6);
+
+    assert_lint_err::<NoUnusedVars>("var a = function() { a(); };", 4);
+    assert_lint_err::<NoUnusedVars>(
+      "var a = function(){ return function() { a(); } };",
+      4,
+    );
+    assert_lint_err::<NoUnusedVars>("const a = () => { a(); };", 6);
+    assert_lint_err::<NoUnusedVars>("const a = () => () => { a(); };", 6);
+  }
+
+  #[test]
+  #[ignore = "pure method analysis is not implemented yet"]
```

Comment:
ditto

Reply:
It can be done in future when some more scope / control flow analysis is implemented, but not now.

Also, I think it's ok to exclude this from the initial version.